### PR TITLE
Add 2022-03 decorators version (stage 3)

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | `Fixes #1, Fixes #2` 
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

This PR adds a new version of the decorators plugin, representing the stage 3 proposal that reached consensus in the March 2022 TC39 meeting: https://github.com/babel/babel/pull/14836#issuecomment-1208270105

Also, the `decoratorsBeforeExport` option is disallowed with the new version and decorators are always after `export` (this aligns with the Babel 8 behavior).

:information_source: The first two commits only copy the 2021-12 files to 2022-03. I suggest hiding it when reviewing (https://github.com/babel/babel/pull/14836/files/915df0ba082dd55dfe5ddc0abf0967ac2ecb571c..8dd440618dbd1fd75fe467b272ccb1859f8c7865), to see the diff between the 2022-03 code/tests and the 2021-12 code/tests.

cc @pzuraq @jkrems

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/14836"></a>

### CODE REVIEW

File: packages/babel-core/src/tools/build-external-helpers.ts [hunk 1]
```diff
@@ -177,7 +177,11 @@ function buildHelpers(
     helpers.ensure(name, File);
     const { nodes } = helpers.get(name, getHelperReference, ref);
 
-    body.push(...nodes);
+    body.push(
+      expressionStatement(
+        callExpression(functionExpression(null, [], blockStatement(nodes)), []),
+      ),
+    );
```
Comment (nicolo-ribaudo) [inline packages/babel-core/src/tools/build-external-helpers.ts:184]: When bundling `babelHelpers`, we used to merge all the functions declarations in a single scope. This doesn't work when multiple helpers have different internal functions with the same name, which is the cause with the new `applyDecs2203` helper and `applyDecs`. This fixes it by wrapping them in IIFEs.

File: packages/babel-helpers/src/helpers/applyDecs2203.js [hunk 2]
```diff
@@ -0,0 +1,797 @@
...
+      break;
+    default:
+      kindStr = "field";
+  }
+
+  var ctx = {
+    kind: kindStr,
+    name: isPrivate ? "#" + name : name,
+    static: isStatic,
+    private: isPrivate,
```
Comment (nicolo-ribaudo) [inline packages/babel-helpers/src/helpers/applyDecs2203.js:168]: These two lines are the only changes compared to the old helper.

**Review by jkrems — COMMENTED** [review]
Only skimmed and left a quick question - thanks for working on this!

File: packages/babel-plugin-proposal-decorators/test/fixtures/2022-03-accessors--to-es2015/private/exec.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+expect(bContext.access.get.call(foo)).toBe(124);
+bContext.access.set.call(foo, 123);
+expect(bContext.access.get.call(foo)).toBe(125);
+expect(bContext.name).toBe('#b');
+expect(bContext.kind).toBe('accessor');
+expect(bContext.static).toBe(false);
+expect(bContext.private).toBe(true);
+expect(typeof bContext.addInitializer).toBe('function');
+expect(typeof bContext.setMetadata).toBe('function');
```
[the remark this answers is not in the document]
Comment (jkrems) [reply packages/babel-plugin-proposal-decorators/test/fixtures/2022-03-accessors--to-es2015/private/exec.js:53]: I'll let @pzuraq take that. I've been only following the proposal progress on the sidelines. AFAIK the `is*` renaming and metadata split were the two notable changes. But it looks like there also might have been a unification of `access`..? https://github.com/tc39/proposal-decorators/commit/13e07b31e01afd0b52892bd4acd1385c344a9880

[the remark this answers is not in the document]
Comment (pzuraq) [reply packages/babel-plugin-proposal-decorators/test/fixtures/2022-03-accessors--to-es2015/private/exec.js:53]: Yes, unification of `access` was one other change we made. I'll comment with a list of all changes

**Comment by pzuraq** [comment]
Thanks for pushing this forward @nicolo-ribaudo! I've been meaning to and just haven't had time.

Here are all the major changes we should make sure to get in this round:

1. `initialize` has been renamed to `init` (this may have already been addressed)
2. `isPrivate` and `isStatic` renamed to `private` and `static`
3. Metadata has been removed
4. `access` is available for all public and private class elements
5. `@(expr)()` is no longer allowed, must wrap the whole expression in parens

I _believe_ that's everything, reviewing the repo issues and PRs. I haven't had a chance to dig back into the actual spec changes, I'll try to find some time this week to review those and make sure nothing slipped between the cracks.

**Comment by nicolo-ribaudo** [comment]
Thanks @pzuraq! I updated the PR.

**Comment by JLHwung** [comment]
It seems to me https://github.com/babel/babel/pull/14666 should be part of this PR, but that one is pending https://github.com/tc39/ecma262/pull/2417#discussion_r896041328. I think we intend to allow `@foo.#bar` as decorator but not `@#bar`. @pzuraq Can you confirm?

File: packages/babel-helpers/src/helpers/applyDecs2203.js [hunk 4]
```diff
@@ -0,0 +1,619 @@
+/* @minVersion 7.17.8 */
```
Comment (JLHwung) [inline packages/babel-helpers/src/helpers/applyDecs2203.js:1]: Should it be `7.19.xx`?

Comment (nicolo-ribaudo) [reply packages/babel-helpers/src/helpers/applyDecs2203.js:1]: It needs to be updated when publishing, some tests fail when helpers specify a future version.

**Comment by nicolo-ribaudo** [comment]
I don't know why the e2e test is failing :confused:

**Comment by nicolo-ribaudo** [comment]
I can reproduce the failure locally downloading the CI artifact

**Comment by nicolo-ribaudo** [comment]
CI is green; please review again! (https://github.com/babel/babel/pull/14836/files/915df0ba082dd55dfe5ddc0abf0967ac2ecb571c..8dd440618dbd1fd75fe467b272ccb1859f8c7865)

**Comment by JLHwung** [comment]
Since https://github.com/babel/babel/pull/14353 is merged and shipped in decorator version `2021-12` and I assume that should have been a `2022-03` thing, we may want to clarify the differences between `2021-12` and `2022-03` in docs:

- `2021-12` <7.17.8 is indeed `2021-12`
- `2021-12` >=7.17.8 is `2021-12` and spec updates covered in #14353
- `2022-03` >=7.19.0 is indeed `2022-03`

**Comment by nicolo-ribaudo** [comment]
I would word it as:
- decorators valid according to the proposal as it was presented on 2021-12 can be run using `version: 2021-12`
- decorators valid according to the proposal as it was presented on 2022-03 can be run using `version: 2021-03`

**Comment by nicolo-ribaudo** [comment]
> And also metadata.initialize is changed to metadata.init.

The 2021-12 helper works with both to avoid breaking: it first checks `.init`, and fallbacks to `.initialize`

File: packages/babel-parser/src/parser/statement.ts [hunk 5]
```diff
@@ -580,6 +580,14 @@ export default abstract class StatementParser extends ExpressionParser {
...
         this.expect(tt.parenR);
         expr = this.wrapParenthesis(startPos, startLoc, expr);
+
+        if (
+          this.getPluginOption("decorators", "allowCallParenthesized") !== false
+        ) {
+          node.expression = this.parseMaybeDecoratorArguments(expr);
+        } else {
+          node.expression = expr;
+        }
```
Comment (JLHwung) [inline packages/babel-parser/src/parser/statement.ts:590]: Can we always parse decorator arguments and throw a recoverable error when `allowCallParenthesized` is `false`?

**Comment by JLHwung** [comment]
Not if one enabled `createParenthesizedExpressions`:

https://github.com/babel/babel/blob/14c3f1215ded1aa18d2c5b0baca14f56162222da/packages/babel-generator/test/fixtures/decorators/decorator-parenthesized-expression-createParenthesizedExpression/output.js#L13

**Review by liuxingbaoyu — APPROVED** [review]

**Pull request merged**